core/dnsserver/watch.go

```diff
@@ -0,0 +1,153 @@
+package dnsserver
+
+import (
+	"io"
+	"log"
+	"sync"
+
+	"github.com/miekg/dns"
+
+	"github.com/coredns/coredns/pb"
+	"github.com/coredns/coredns/plugin"
+	"github.com/coredns/coredns/plugin/pkg/watch"
+)
+
+// watcher contains all the data needed to manage watches
+type watcher struct {
+	changes watch.Chan
+	counter int64
+	watches map[string]watchlist
+	plugins []watch.Watchable
+	mutex   sync.Mutex
+}
+
+type watchlist map[int64]pb.DnsService_WatchServer
+
+func newWatcher(zones map[string]*Config) *watcher {
+	w := &watcher{changes: make(watch.Chan), watches: make(map[string]watchlist)}
+
+	for _, config := range zones {
+		plugins := config.Handlers()
+		for _, p := range plugins {
+			if x, ok := p.(watch.Watchable); ok {
+				x.SetWatchChan(w.changes)
+				w.plugins = append(w.plugins, x)
+			}
+		}
+	}
+
+	//TODO: maybe a stop channel, work properly with reloads?
+	go w.processWatches()
+	return w
+}
+
+func (w *watcher) nextID() int64 {
+	w.mutex.Lock()
+
+	w.counter++
+	id := w.counter
+
+	w.mutex.Unlock()
+	return id
+}
+
+// watch is used to monitor the results of a given query. CoreDNS will push updated
+// query responses down the stream.
+func (w *watcher) watch(stream pb.DnsService_WatchServer) error {
+	for {
+		in, err := stream.Recv()
+		if err == io.EOF {
+			return nil
+		}
+		if err != nil {
+			return err
+		}
+		create := in.GetCreateRequest()
+		if create != nil {
+			msg := new(dns.Msg)
+			err := msg.Unpack(create.Query.Msg)
+			if err != nil {
+				// TODO: should write back an error response not break the stream
```

Comment:
should we have some sort of error message we can give back when we say that we couldn't create it?

Reply:
OK - I guess we can log an error to the coredns log.  I didn't se a way to send an error message/code to the client.  Or do I need to expand that... OH I see.. thats what the `.proto` file is for...
OK, will add an message field to the WatchResponse definition.

```
string message = 5
```

... and then add the client part to dnsgrpc.go.

or am i still confused.